src/xcode/ENA/ENA/Source/Services/PPAnalyticsCollector/PPAAnalyticsSubmissionCollector.swift

```diff
@@ -0,0 +1,220 @@
+////
+// 🦠 Corona-Warn-App
+//
+
+import Foundation
+
+final class PPAAnalyticsSubmissionCollector {
+
+	init(
+		store: Store,
+		coronaTestService: CoronaTestService
+	) {
+		// We put the PPAnalyticsData protocol and its implementation in a seperate file because this protocol is only used by the collector. And only the collector should use it!
+		// This way we avoid the direct access of analytics data at other places over the store.
+		guard let store = store as? (Store & PPAnalyticsData) else {
+			Log.error("I will never submit any analytics data. Could not cast to correct store protocol", log: .ppa)
+			fatalError("I will never submit any analytics data. Could not cast to correct store protocol")
+		}
+
+		self.store = store
+		self.coronaTestService = coronaTestService
+	}
+
+	// MARK: - Internal
+
+	// swiftlint:disable:next cyclomatic_complexity
+	func logKeySubmissionMetadata(_ keySubmissionMetadata: PPAKeySubmissionMetadata) {
+		switch keySubmissionMetadata {
+		case let .create(metadata, type):
+			switch type {
+			case .pcr:
+				store.keySubmissionMetadata = metadata
+			case .antigen:
+				store.antigenKeySubmissionMetadata = metadata
+			}
+		case let .submitted(submitted, type):
+			switch type {
+			case .pcr:
+				store.keySubmissionMetadata?.submitted = submitted
+			case .antigen:
+				store.antigenKeySubmissionMetadata?.submitted = submitted
+			}
+		case let .submittedInBackground(inBackground, type):
+			switch type {
+			case .pcr:
+				store.keySubmissionMetadata?.submittedInBackground = inBackground
+			case .antigen:
+				store.antigenKeySubmissionMetadata?.submittedInBackground = inBackground
+			}
+		case let .submittedAfterCancel(afterCancel, type):
+			switch type {
+			case .pcr:
+				store.keySubmissionMetadata?.submittedAfterCancel = afterCancel
+			case .antigen:
+				store.antigenKeySubmissionMetadata?.submittedAfterCancel = afterCancel
+			}
+		case let .submittedAfterSymptomFlow(afterSymptomFlow, type):
+			switch type {
+			case .pcr:
+				store.keySubmissionMetadata?.submittedAfterSymptomFlow = afterSymptomFlow
+			case .antigen:
+				store.antigenKeySubmissionMetadata?.submittedAfterSymptomFlow = afterSymptomFlow
+			}
+		case let .submittedWithTeletan(withTeletan, type):
+			switch type {
+			case .pcr:
+				store.keySubmissionMetadata?.submittedWithTeleTAN = withTeletan
+			case .antigen:
+				store.antigenKeySubmissionMetadata?.submittedWithTeleTAN = withTeletan
+			}
+		case let .lastSubmissionFlowScreen(flowScreen, type):
+			switch type {
+			case .pcr:
+				store.keySubmissionMetadata?.lastSubmissionFlowScreen = flowScreen
+			case .antigen:
+				store.antigenKeySubmissionMetadata?.lastSubmissionFlowScreen = flowScreen
+			}
+		case let .advancedConsentGiven(advanceConsent, type):
+			switch type {
+			case .pcr:
+				// this is as per techspecs, this value is false in case TAN submission
+				if store.keySubmissionMetadata?.submittedWithTeleTAN == false && advanceConsent == true {
```

Comment:
```suggestion
				if store.keySubmissionMetadata?.submittedWithTeleTAN == false {
```

I think it would be easier to read like that, I had to think about it quite a lot to understand why this works. Same for antigen below.

Reply:
Changed